File: factoryMethod/factory_method.cpp
##include<stdout>

class SplitterFactory {
public:
    virtual ISplitter* createSplitter()=0;
    virtual ~SplitterFactory(){}
};

//文件分割器
class ISplitter {
public:
    virtual void split()=0;
    virtual ~ISplitter(){}
};

// 具体的实现类：1
class BinarySplitter: public ISplitter {
public:
    virtual void split(){
        std::out<<"split binary"<<std::endl;
    }
};

// 具体的实现类：2
class TextSplitter: public ISplitter {
public:
    virtual void split(){
        std::out<<"text binary"<<std::endl;
    }
};

//具体工厂类
class BinarySplitterFactory: public SplitterFactory {
public:
    virtual ISplitter* createSplitter() {
        return new(BinarySplitter);
    }
};

class TextSplitterFactory: public SplitterFactory {
public:
    virtual ISplitter* createSplitter() {
        return new(TextSplitter);
    }
};

class MainForm {
    SplitterFactory* factory;
public:
    MainForm(SplitterFactory* factory):factory(factory){}

    void Do(){
        ISplitter* splitter = this->factory.createSplitter();
        splitter->split();
    }
};

File: factoryMethod/factoryMethod.go
package factoryMethod

import "fmt"

// 工厂方法的接口
type ISplitterFactory interface {
	CreateSplitter() ISplitter
}

// 文件分割器接口
type ISplitter interface {
	Split()
}

// concrete splitter:1
type BinarySplitter struct{}

func (s *BinarySplitter) Split() {
	fmt.Println("split binary")
}

// concrete splitter:2
type TextSplitter struct{}

func (s *TextSplitter) Split() {
	fmt.Println("text binary")
}

//每个类都有一个对应的工厂类
type BinarySplitterFactory struct{}

func (c *BinarySplitterFactory) CreateSplitter() ISplitter {
	return &BinarySplitter{}
}

type TextSplitterFactory struct{}

func (c *TextSplitterFactory) CreateSplitter() ISplitter {
	return &TextSplitter{}
}
